mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,323 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::task::{Spawn, SpawnExt};
+use tracing::{debug, warn};
+
+use crate::object::ObjectId;
+
+use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheResult};
+
+pub struct MultilevelDataCache<DiskCache, ExpressCache, Runtime> {
+    disk_cache: Arc<DiskCache>,
+    express_cache: ExpressCache,
+    runtime: Runtime,
+}
+
+impl<DiskCache: DataCache, ExpressCache: DataCache, Runtime: Spawn>
+    MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    pub fn new(disk_cache: Arc<DiskCache>, express_cache: ExpressCache, runtime: Runtime) -> Self {
+        assert_eq!(disk_cache.block_size(), express_cache.block_size());
+        Self {
+            disk_cache,
+            express_cache,
+            runtime,
+        }
+    }
+}
+
+#[async_trait]
+impl<DiskCache: DataCache + Sync + Send + 'static, ExpressCache: DataCache + Sync, Runtime: Spawn + Sync> DataCache
+    for MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    async fn get_block(
+        &self,
+        cache_key: &ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+        object_size: usize,
+    ) -> DataCacheResult<Option<ChecksummedBytes>> {
+        if let Some(data) = self
+            .disk_cache
+            .get_block(cache_key, block_idx, block_offset, object_size)
+            .await?
```

Comment:
Error handling requires a bit more thought: if there is an error here, we probably want to try the shared cache rather than bailing out.